source/common/grpc/google_async_client_impl.cc

```diff
@@ -205,7 +205,8 @@ void GoogleAsyncStreamImpl::resetStream() {
 }
 
 void GoogleAsyncStreamImpl::writeQueued() {
-  if (!call_initialized_ || finish_pending_ || write_pending_ || write_pending_queue_.empty()) {
+  if (!call_initialized_ || finish_pending_ || write_pending_ || write_pending_queue_.empty() ||
```

Comment:
This seems a legitimate change; we shouldn't be queueing new writes if we're draining for sure. Does this solve your teardown race for the Envoy gRPC client as well?

Trying to recall our earlier conversation, it seems we should get a clear teardown ordering. We should ensure that all the gRPC consumers are properly removed before we begin the TLS teardown.

Reply:
Yes, it solves the test flakiness.  tested with runs_per_test=1000.

Reply:
We could do that teardown ordering in a separate PR.  We need to get this in to unblock CI tests.

Reply:
OK, I will merge given that you folks have ownership on the underlying issue. Thanks.